### PR TITLE
feat(client): replace jersey with okhttp in hugegraph-client (Draft PR)

### PR DESCRIPTION
**Pull request by zhenyuT**
- [https://github.com/apache/incubator-hugegraph-toolchain/issues/505](https://github.com/apache/incubator-hugegraph-toolchain/issues/505)

1、直接修改hugegraph-client代码，覆盖common中的类
2、对于https自定义证书的实现暂时缺失
3、运行ci失败，且耗时较久(30min)，部分单测执行出现connection timeout的异常。推测可能okhttp client的close存在问题，多次创建client实例导致内存/连接泄漏，导致存在性能问题

麻烦先看下代码，如果大体方向没问题的话，我再继续进行代码优化 (以及调整 hugegrpah-common 中的依赖, 目前是直接在 client 调整)

### CODE REVIEW

**Comment by imbajin** [comment]
thanks and some todos：
- [ ] seems also need handle https test in ci（loader/hubble..）？
- [ ] fix dependencies check as the contribution doc example in website

File: hugegraph-client/src/test/java/org/apache/hugegraph/BaseClientTest.java [hunk 1]
```diff
@@ -49,7 +49,8 @@
 
 public class BaseClientTest {
 
-    protected static final String BASE_URL = "http://127.0.0.1:8080";
+//    protected static final String BASE_URL = "http://127.0.0.1:8080";
+    protected static final String BASE_URL = "http://172.19.102.150:8080";
```
Comment (simon824) [inline hugegraph-client/src/test/java/org/apache/hugegraph/BaseClientTest.java:53]: Don't commit these changes

**Review by simon824 — COMMENTED** [review]
Overall looks good, but we need fix ci first and improve codestyle refer to https://github.com/apache/incubator-hugegraph/blob/master/hugegraph-style.xml

**Comment by zhenyuT** [comment]
> Overall looks good, but we need fix ci first and improve codestyle refer to https://github.com/apache/incubator-hugegraph/blob/master/hugegraph-style.xml

I've done my best to fix the issues mentioned on,but there is one more problem. When I add the licence header in the file "hugegraph-client/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker",the unit test will run error.

File: hugegraph-client/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 2]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (imbajin) [inline hugegraph-client/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: And is this file necessary? (maybe we could try to add license header first)

```suggestion
# license header in shell
# ....
# end
mock-maker-inline
```

Comment (zhenyuT) [reply hugegraph-client/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: > maybe we could try to **add license header** for it first if (it's) necessary

This just works fine, thank you.

File: hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/GraphAPI.java [hunk 3]
```diff
@@ -76,10 +73,11 @@ public static String formatProperties(Map<String, Object> properties) {
...
-        return encode(json);
+//        return encode(json);
+        return json;
     }
 
-    public static String encode(String raw) {
-        return UriComponent.encode(raw, Type.QUERY_PARAM_SPACE_ENCODED);
-    }
+//    public static String encode(String raw) {
+//        return UriComponent.encode(raw, Type.QUERY_PARAM_SPACE_ENCODED);
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/GraphAPI.java:81]: any alternative plan?

File: hugegraph-client/src/test/java/org/apache/hugegraph/unit/RestResultTest.java [hunk 4]
```diff
@@ -55,7 +56,7 @@
 
 public class RestResultTest extends BaseUnitTest {
 
-    private jakarta.ws.rs.core.Response mockResponse;
+    private okhttp3.Response mockResponse;
```
Comment (javeme) [inline hugegraph-client/src/test/java/org/apache/hugegraph/unit/RestResultTest.java:59]: just import the Response class?

**Review by javeme — COMMENTED** [review]
Thanks for your contribution

**Comment by javeme** [comment]
> 麻烦先看下代码，如果大体方向没问题的话，我再继续进行代码优化 (以及调整 hugegrpah-common 中的依赖, 目前是直接在 client 调整)

整体方案没问题的。
CI失败，看了下是超时问题SocketTimeoutException，请确保各类超时时间设置正确了：
```
MultiNodeShortestPathApiTest Time elapsed: 10.038 s  <<< ERROR!
java.net.SocketTimeoutException: timeout
```

**Comment by zhenyuT** [comment]
> > 麻烦先看下代码，如果大体方向没问题的话，我再继续进行代码优化 (以及调整 hugegrpah-common 中的依赖, 目前是直接在 client 调整)
> 
> 整体方案没问题的。 CI失败，看了下是超时问题SocketTimeoutException，请确保各类超时时间设置正确了：
> 
> ```
> MultiNodeShortestPathApiTest Time elapsed: 10.038 s  <<< ERROR!
> java.net.SocketTimeoutException: timeout
> ```

有点奇怪，我fork到自己仓库，执行是成功的，没有遇到超时的问题。两边执行是有什么区别吗？

**Comment by z7658329** [comment]
seems good,  do you have test introducing the new hugegrap-client dependency in a spring boot project and ensured that it starts up correctly? and you need to cover two scenarios: jdk8/jdk11 + springboot2.X and jdk17 + springboot3,   you can refer [495](https://github.com/apache/incubator-hugegraph-toolchain/issues/495)  @zhenyuT

**Comment by zhenyuT** [comment]
> seems good, do you have test introducing the new hugegrap-client dependency in a spring boot project and ensured that it starts up correctly? and you need to cover two scenarios: jdk8/jdk11 + springboot2.X and jdk17 + springboot3, you can refer [495](https://github.com/apache/incubator-hugegraph-toolchain/issues/495) @zhenyuT

I've test with springboot2.X and springboot3,but just simple test.

File: hugegraph-client/src/test/java/org/apache/hugegraph/unit/RestResultTest.java (hunk 4 above)
Comment (zhenyuT) [reply hugegraph-client/src/test/java/org/apache/hugegraph/unit/RestResultTest.java:59]: > 

这里之前用的jakartar包中的Response，现在实现改为okhttp，因此对应mock的对象也修改为okhttp包中的Response。这样做是存在什么问题？

File: hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/GraphAPI.java (hunk 3 above)
Comment (zhenyuT) [reply hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/GraphAPI.java:81]: > any alternative plan?

看原来代码的意图，是要对url中的参数进行编码。但是okhttp中使用HttpUrl.Builder构建url，其实会自动进行编码的，所以感觉没必要由使用方对参数编码后再传参

File: hugegraph-client/pom.xml [hunk 5]
```diff
@@ -37,28 +37,73 @@
     </description>
 
     <properties>
+        <lombok.version>1.18.8</lombok.version>
+        <okhttp.version>4.10.0</okhttp.version>
```
Comment (z7658329) [inline hugegraph-client/pom.xml:41]: shoule extract dependency version  to  root pom.xml， sub module just import?

Comment (imbajin) [reply hugegraph-client/pom.xml:41]: > shoule extract dependency version to root pom.xml， sub module just import?

seems no need, only extract dependencies that **multiple modules** need to use (leave single usage alone?)

**Comment by simon824** [comment]
LGTM, @zhenyuT please resolve conflicts when you are free.

File: hugegraph-client/src/main/java/org/apache/hugegraph/api/auth/UserAPI.java [hunk 6]
```diff
@@ -40,32 +40,33 @@ protected String type() {
...
 
     public User get(Object id) {
-        RestResult result = this.client.get(this.path(), formatEntityId(id));
+        OkhttpRestResult result = this.client.get(this.path(), formatEntityId(id));
         return result.readObject(User.class);
     }
 
     public UserRole getUserRole(Object id) {
-        String idEncoded = RestClient.encode(formatEntityId(id));
+//        String idEncoded = RestClient.encode(formatEntityId(id));
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/api/auth/UserAPI.java:53]: can we remove the unused code

File: hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/EdgeAPI.java [hunk 7]
```diff
@@ -46,16 +45,17 @@ protected String type() {
...
-        RestResult result = this.client.post(this.path(), edge);
+        OkhttpRestResult result = this.client.post(this.path(), edge);
         return result.readObject(Edge.class);
     }
 
     public List<String> create(List<Edge> edges, boolean checkVertex) {
-        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
-        headers.putSingle("Content-Encoding", BATCH_ENCODING);
+//        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
+//        headers.putSingle("Content-Encoding", BATCH_ENCODING);
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/EdgeAPI.java:54]: ditto

File: hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/EdgeAPI.java [hunk 8]
```diff
@@ -68,29 +68,30 @@ public List<String> create(List<Edge> edges, boolean checkVertex) {
 
     public List<Edge> update(BatchEdgeRequest request) {
         this.client.checkApiVersion("0.45", "batch property update");
-        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
-        headers.putSingle("Content-Encoding", BATCH_ENCODING);
-        RestResult result = this.client.put(this.batchPath(), null,
+//        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
+//        headers.putSingle("Content-Encoding", BATCH_ENCODING);
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/EdgeAPI.java:72]: ditto

File: hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/EdgeAPI.java [hunk 9]
```diff
@@ -68,29 +68,30 @@ public List<String> create(List<Edge> edges, boolean checkVertex) {
...
+        Headers headers = new Headers.Builder().add("Content-Encoding", BATCH_ENCODING).build();
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/EdgeAPI.java:73]: define a const?

File: hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/VertexAPI.java [hunk 10]
```diff
@@ -68,19 +66,21 @@ public List<Object> create(List<Vertex> vertices) {
...
                                             request, headers);
         return result.readList(this.type(), Vertex.class);
     }
 
     public int update(BatchOlapPropertyRequest request) {
         this.client.checkApiVersion("0.59", "olap property batch update");
-        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
-        headers.putSingle("Content-Encoding", BATCH_ENCODING);
+//        MultivaluedHashMap<String, Object> headers = new MultivaluedHashMap<>();
+//        headers.putSingle("Content-Encoding", BATCH_ENCODING);
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/api/graph/VertexAPI.java:80]: ditto

File: hugegraph-client/src/main/java/org/apache/hugegraph/driver/HugeClient.java [hunk 11]
```diff
@@ -60,7 +60,8 @@ public HugeClient(HugeClientBuilder builder) {
                                          builder.maxConnsPerRoute(),
                                          builder.trustStoreFile(),
                                          builder.trustStorePassword());
-        } catch (ProcessingException e) {
+        } catch (Exception e) {
+            log.error("", e);
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/driver/HugeClient.java:64]: please improve the log message

File: hugegraph-client/src/main/java/org/apache/hugegraph/rest/OkhttpAbstractRestClient.java [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+                    .readTimeout(okhttpConfig.getTimeout(), TimeUnit.MILLISECONDS);
+        }
+
+        if(okhttpConfig.getIdleTime()!=null) {
+            ConnectionPool connectionPool = new ConnectionPool(5, okhttpConfig.getIdleTime(), TimeUnit.MILLISECONDS);
+            builder.connectionPool(connectionPool);
+        }
+
+
+        //auth
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/rest/OkhttpAbstractRestClient.java:148]: can we add more details for comments

File: hugegraph-client/src/main/java/org/apache/hugegraph/rest/OkhttpConfig.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+package org.apache.hugegraph.rest;
+
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Setter;
+
+@Builder
+@Getter
+@Setter
+public class OkhttpConfig {
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/rest/OkhttpConfig.java:27]: expect a blank line after the class define

File: hugegraph-client/src/main/java/org/apache/hugegraph/rest/OkhttpRestClient.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+ * under the License.
+ */
+
+package org.apache.hugegraph.rest;
+
+import okhttp3.Headers;
+
+import java.util.Map;
+
+public interface OkhttpRestClient {
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/rest/OkhttpRestClient.java:24]: don't need to Okhttp prefix, we expect to keep the same interface name

File: hugegraph-client/src/main/java/org/apache/hugegraph/rest/OkhttpRestResult.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.SneakyThrows;
+import okhttp3.Headers;
+import okhttp3.Response;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class OkhttpRestResult {
```
Comment (javeme) [inline hugegraph-client/src/main/java/org/apache/hugegraph/rest/OkhttpRestResult.java:32]: don't need to Okhttp prefix

File: hugegraph-client/src/test/java/org/apache/hugegraph/unit/RestResultTest.java (hunk 4 above)
Comment (javeme) [reply hugegraph-client/src/test/java/org/apache/hugegraph/unit/RestResultTest.java:59]: ok, I mean we can try to import okhttp3.Response

File: pom.xml [hunk 16]
```diff
@@ -151,17 +151,35 @@
...
-                <type>pom</type>
-                <scope>import</scope>
-            </dependency>
+<!--            <dependency>-->
+<!--                <groupId>org.glassfish.jersey</groupId>-->
+<!--                <artifactId>jersey-bom</artifactId>-->
+<!--                <version>${jersey.version}</version>-->
+<!--                <type>pom</type>-->
+<!--                <scope>import</scope>-->
+<!--            </dependency>-->
```
Comment (javeme) [inline pom.xml:160]: we can remove all the unused code

**Comment by javeme** [comment]
implemented by https://github.com/apache/incubator-hugegraph-commons/pull/133
